### PR TITLE
LEDs (disable patterns, framework for amplification and note signaling)

### PR DESCRIPTION
**Pull request by Malachi-lewis**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+          10 - (int) amplifierTimer.get(), isRedAlliance ? RGB.BLUE : RGB.RED);
+      if (amplifierTimer.get() > 10) {
+        isOpponentAmplified = false;
+        amplifierTimer.reset();
+        startSignal(Constants.AMPLIFICATION_RUMBLE_TIME, Constants.AMPLIFICATION_RUMBLE_INTENSITY);
+      }
+    }
+  }
+
+  public void startAmplification(boolean isOpposingTeam) {
```
Comment (lohengrin332) [inline src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java:99]: All function parameters and variables should be declared `final` wherever possible.

File: src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    this.driverRumbleConsumer = driverRumbleConsumer;
+
+    tick = 0;
+    patternTick = 0;
+    amplifierTimer = new Timer();
+    rumbleTimer = new Timer();
+    isAllianceAmplified = false;
+    isOpponentAmplified = false;
+
+    this.isRedAlliance = DriverStation.getAlliance().equals(DriverStation.Alliance.Red);
```
Comment (azvanderpas) [inline src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java:49]:
```suggestion
    this.isRedAlliance = DriverStation.getAlliance().get().equals(DriverStation.Alliance.Red);
```
getAlliance() returns an Optional<DriverStation.Allance>, so get() is needed to get the contained object.

Comment (Malachi-lewis) [reply src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java:49]: Fixed

**Review by lohengrin332 — DISMISSED** [review]

File: src/main/java/frc/robot/subsystems/color/patterns/BIFlag.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package frc.robot.subsystems.color.patterns;
+
+import frc.robot.subsystems.color.RGB;
+
+public class BIFlag {
```
Comment (azvanderpas) [inline src/main/java/frc/robot/subsystems/color/patterns/BIFlag.java:5]: May as well make the pattern classes static, right?

File: src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    ledBuffer = new AddressableLEDBuffer(numberOfLEDS);
+    ledStrip.setLength(numberOfLEDS);
+    ledStrip.setData(ledBuffer);
+    ledStrip.start();
+
+    // Setup Disable Patterns
+    TransFlag.numberOfLEDS = numberOfLEDS;
+    FireFlyPattern.numberOfLEDS = numberOfLEDS;
+    BIFlag.numberOfLEDS = numberOfLEDS;
+    RainbowPattern.numberOfLEDS = numberOfLEDS;
```
Comment (azvanderpas) [inline src/main/java/frc/robot/subsystems/color/SignalingSubsystem.java:62]: Is there a good reason not to just have the patterns use the Constants.LED_COUNT directly? This is an unusual way to set values of another class.

File: src/main/java/frc/robot/subsystems/color/patterns/BIFlag.java (hunk 3 above)
Comment (Malachi-lewis) [reply src/main/java/frc/robot/subsystems/color/patterns/BIFlag.java:6]: You can't sadly.

**Review by lohengrin332 — APPROVED** [review]

**Pull request merged**